### PR TITLE
Enable file mode="x" with a single_file=True for dask dataframe to_csv

### PR DESCRIPTION
**Pull request by GenevieveBuckley**
- [x] Closes https://github.com/dask/dask/issues/10442
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

Issue https://github.com/dask/dask/issues/10442 describes two closely related issues.
This PR addresses the second issue, using file `mode="x"` with a `single_file=True` in dask dataframe `to_csv()`

@hendrikmakait provides a test demonstrating the problem in the issue, and points out the problematic line here:
> (See https://github.com/dask/dask/pull/10441#discussion_r1293437311 for a pointer to where the problem might be.) cc @benrutter

### CODE REVIEW

**Comment by benrutter** [comment]
Hey @GenevieveBuckley - this looks awesome! I think (am not sure) that the replace *might* need to move a few lines down?

My understandimg is, the part of the code being edited runs if the user has marked single_file, and dask is changing the write mode because there's a disconnect between the user making a single call, and dask is a call for each partition.

If that's the case, and the user marks "x" (no rewrite), they'd probably want some guarantees on the operation crashing out if the file already existed, which they'll miss if "x" is removed before any write operation.

Does the function need to check the file doesn't initially exist?

Also, I'm fairly new to dask, so please do take what I say with a pinch of salt- I could definitely be wrong 😅

**Comment by GenevieveBuckley** [comment]
I assumed that that was taken care of by mode="x" still being in effect when the first file partition is written to disk two lines above this change.
```python
first_file = open_file(filename, mode=mode, **file_options)
```

But I did not actually write a test for that, so thank you for mentioning it

**Comment by GenevieveBuckley** [comment]
Update: yes it's fine, works as expected. I've added an extra test now.

```python
In [1]: import pandas as pd

In [2]: import dask.dataframe as dd

In [3]: df0 = pd.DataFrame(
   ...:     {"x": ["a", "b", "c", "d"], "y": [1, 2, 3, 4]}
   ...: )
   ...: df = dd.from_pandas(df0, npartitions=2)

In [4]: df.to_csv("csv_path.csv", index=False, mode="x", single_file=True)
Out[4]: ['/Users/genevieb/Documents/temp/csv_path.csv']

In [5]: # Try creating the same file again, should error because "x" mode needs the file to NOT exist already

In [6]: df.to_csv("csv_path.csv", index=False, mode="x", single_file=True)
---------------------------------------------------------------------------
FileExistsError                           Traceback (most recent call last)
Cell In[6], line 1
----> 1 df.to_csv("csv_path.csv", index=False, mode="x", single_file=True)

File ~/Documents/GitHub/dask/dask/dataframe/io/csv.py:994, in to_csv(df, filename, single_file, encoding, mode, name_function, compression, compute, scheduler, storage_options, header_first_partition_only, compute_kwargs, **kwargs)
    990         compute_kwargs["scheduler"] = scheduler
    992     import dask
--> 994     return list(dask.compute(*values, **compute_kwargs))
    995 else:
    996     return values

File ~/Documents/GitHub/dask/dask/threaded.py:89, in get(dsk, keys, cache, num_workers, pool, **kwargs)
     86     elif isinstance(pool, multiprocessing.pool.Pool):
     87         pool = MultiprocessingPoolExecutor(pool)
---> 89 results = get_async(
     90     pool.submit,
     91     pool._max_workers,
     92     dsk,
     93     keys,
     94     cache=cache,
     95     get_id=_thread_get_id,
     96     pack_exception=pack_exception,
     97     **kwargs,
     98 )
    100 # Cleanup pools associated to dead threads
    101 with pools_lock:

File ~/Documents/GitHub/dask/dask/local.py:511, in get_async(submit, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, chunksize, **kwargs)
    509         _execute_task(task, data)  # Re-execute locally
    510     else:
--> 511         raise_exception(exc, tb)
    512 res, worker_id = loads(res_info)
    513 state["cache"][key] = res

File ~/Documents/GitHub/dask/dask/local.py:319, in reraise(exc, tb)
    317 if exc.__traceback__ is not tb:
    318     raise exc.with_traceback(tb)
--> 319 raise exc

File ~/Documents/GitHub/dask/dask/local.py:224, in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    222 try:
    223     task, data = loads(task_info)
--> 224     result = _execute_task(task, data)
    225     id = get_id()
    226     result = dumps((result, id))

File ~/Documents/GitHub/dask/dask/dataframe/io/csv.py:792, in _write_csv(df, fil, depend_on, **kwargs)
    791 def _write_csv(df, fil, *, depend_on=None, **kwargs):
--> 792     with fil as f:
    793         df.to_csv(f, **kwargs)
    794     return os.path.normpath(fil.path)

File ~/mambaforge/envs/dask-dev/lib/python3.10/site-packages/fsspec/core.py:102, in OpenFile.__enter__(self)
     99 def __enter__(self):
    100     mode = self.mode.replace("t", "").replace("b", "") + "b"
--> 102     f = self.fs.open(self.path, mode=mode)
    104     self.fobjects = [f]
    106     if self.compression is not None:

File ~/mambaforge/envs/dask-dev/lib/python3.10/site-packages/fsspec/spec.py:1241, in AbstractFileSystem.open(self, path, mode, block_size, cache_options, compression, **kwargs)
   1239 else:
   1240     ac = kwargs.pop("autocommit", not self._intrans)
-> 1241     f = self._open(
   1242         path,
   1243         mode=mode,
   1244         block_size=block_size,
   1245         autocommit=ac,
   1246         cache_options=cache_options,
   1247         **kwargs,
   1248     )
   1249     if compression is not None:
   1250         from fsspec.compression import compr

File ~/mambaforge/envs/dask-dev/lib/python3.10/site-packages/fsspec/implementations/local.py:184, in LocalFileSystem._open(self, path, mode, block_size, **kwargs)
    182 if self.auto_mkdir and "w" in mode:
    183     self.makedirs(self._parent(path), exist_ok=True)
--> 184 return LocalFileOpener(path, mode, fs=self, **kwargs)

File ~/mambaforge/envs/dask-dev/lib/python3.10/site-packages/fsspec/implementations/local.py:315, in LocalFileOpener.__init__(self, path, mode, autocommit, fs, compression, **kwargs)
    313 self.compression = get_compression(path, compression)
    314 self.blocksize = io.DEFAULT_BUFFER_SIZE
--> 315 self._open()

File ~/mambaforge/envs/dask-dev/lib/python3.10/site-packages/fsspec/implementations/local.py:320, in LocalFileOpener._open(self)
    318 if self.f is None or self.f.closed:
    319     if self.autocommit or "w" not in self.mode:
--> 320         self.f = open(self.path, mode=self.mode)
    321         if self.compression:
    322             compress = compr[self.compression]

FileExistsError: [Errno 17] File exists: '/Users/genevieb/Documents/temp/csv_path.csv'
```

**Comment by benrutter** [comment]
Sorry @GenevieveBuckley - i'd completely missed that open file call already made! Nice one 😁

**Review by hendrikmakait — APPROVED** [review]
Thanks, @GenevieveBuckley!

File: dask/dataframe/io/csv.py [hunk 1]
```diff
@@ -942,7 +942,7 @@ def to_csv(
         first_file = open_file(filename, mode=mode, **file_options)
         value = to_csv_chunk(dfs[0], first_file, **kwargs)
         append_mode = mode if "a" in mode else mode + "a"
-        append_mode = append_mode.replace("w", "")
+        append_mode = mode.replace("w", "").replace("x", "")
```
Comment (hendrikmakait) [inline dask/dataframe/io/csv.py:945]: Sorry, I messed up the merge with `main`.
```suggestion
        append_mode = append_mode.replace("w", "").replace("x", "")
```

**Pull request merged**